### PR TITLE
Add Custom Model Type

### PR DESCRIPTION
**Pull request by tiopramayudi**
**What this PR does / why we need it**:

This PR is enabled users to create their own predictor implementation. Users just need to specify their docker image. Docker image must be web service that has several endpoints that must be implemented. Several modifications in this PR:
* Update version patch API
* Update all version API's response so it contains custom predictor object
* Update SDK to log new custom model
* Update UI, showing container spec for custom predictor on model service panel
**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

Screen Shot 2021-06-10 at 09 20 59

```release-note

```

**Checklist**

- [x] Added unit test, integration, and/or e2e tests
- [x] Tested locally
- [ ] Updated documentation
- [x] Update Swagger spec if the PR introduce API changes
- [x] Regenerated Golang and Python client if the PR introduce API changes

### CODE REVIEW

File: python/sdk/test/integration_test.py [hunk 1]
```diff
@@ -715,3 +715,43 @@ def test_standard_transformer_with_feast(integration_test_url, project_name, use
 
     assert resp.json()["instances"] == exp_resp["instances"]
     merlin.undeploy(v)
+
+@pytest.mark.integration
+def test_custom_model_without_artifact(integration_test_url, project_name, use_google_oauth):
```
Comment (ariefrahmansyah) [inline python/sdk/test/integration_test.py:720]: can we have test for custom model deployment with artifact?

File: api/cluster/templater.go [hunk 2]
```diff
@@ -205,6 +216,43 @@ func createLoggerSpec(loggerURL string, loggerConfig models.LoggerConfig) *kfsv1
 	}
 }
 
+func createCustomPredictorSpec(modelService *models.Service, resources v1.ResourceRequirements) kfsv1alpha2.PredictorSpec {
+	envVars := modelService.EnvVars
+	envVars = append(envVars, models.EnvVar{Name: envPredictorPort, Value: defaultPredictorPort})
+	envVars = append(envVars, models.EnvVar{Name: envPredictorModelName, Value: modelService.Name})
+	envVars = append(envVars, models.EnvVar{Name: envPredictorArtifactLocation, Value: defaultPredictorArtifactLocation})
+	envVars = append(envVars, models.EnvVar{Name: envPredictorStorageURI, Value: utils.CreateModelLocation(modelService.ArtifactURI)})
```
Comment (ariefrahmansyah) [inline api/cluster/templater.go]: As discussed offline, we can omit `STORAGE_URI` if user doesn't specify the artifact URI.

Comment (tiopramayudi) [reply api/cluster/templater.go]: it is already updated here http://custom-wo-artifact-17.integration-test.models.id.d.gods.golabs.io/v1/models/custom-wo-artifact-17:predict

File: python/sdk/test/integration_test.py (hunk 1 above)
Comment (tiopramayudi) [reply python/sdk/test/integration_test.py:720]: I plan to update it after finishing the sample docker model. The sample will contains webservice code, Dockerfile and Notebook. So I think I update it in the next PR

File: api/api/version_endpoints_api.go [hunk 3]
```diff
@@ -197,6 +206,15 @@ func (c *EndpointsController) UpdateEndpoint(r *http.Request, vars map[string]st
 		return NotFound(err.Error())
 	}
 
+	// validate custom predictor
+	if model.Type == models.ModelTypeCustom {
+		err := c.validateCustomPredictor(ctx, version)
+		if err != nil {
+			log.Errorf("error validating custom predictor: %v", err)
```
[the remark this answers is not in the document]
Comment (tiopramayudi) [reply api/api/version_endpoints_api.go:213]: yeah, will update this

File: api/api/version_endpoints_api.go [hunk 4]
```diff
@@ -375,6 +393,17 @@ func (c *EndpointsController) validateTransformer(ctx context.Context, trans *mo
 	return nil
 }
 
+func (c *EndpointsController) validateCustomPredictor(ctx context.Context, version *models.Version) error {
+	customPredictor := version.CustomPredictor
+	if customPredictor == nil {
+		return errors.New("custom predictor must be specified")
+	}
+	if customPredictor.Image == "" {
```
[the remark this answers is not in the document]
Comment (tiopramayudi) [reply api/api/version_endpoints_api.go:401]: sure we can, I'll update this

File: python/sdk/test/integration_test.py (hunk 1 above)
Comment (ariefrahmansyah) [reply python/sdk/test/integration_test.py:720]: Got it 👌🏻

File: api/client/model_custom_predictor.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+ * Generated by: Swagger Codegen (https://github.com/swagger-api/swagger-codegen.git)
+ */
+
+package client
+
+type CustomPredictor struct {
+	Image           string `json:"image,omitempty"`
+	Command         string `json:"command,omitempty"`
+	Args            string `json:"args,omitempty"`
+	IsArtifactExist bool   `json:"is_artifact_exist,omitempty"`
```
Comment (pradithya) [inline api/client/model_custom_predictor.go:16]: Can this be hidden from user `IsArtifactExist` ?

Comment (pradithya) [reply api/client/model_custom_predictor.go:16]: It seems redundant with `ArtifactUri`

File: python/sdk/merlin/model.py [hunk 6]
```diff
@@ -912,6 +915,34 @@ def log_model(self, model_dir=None):
         validate_model_dir(self._model.type, None, model_dir)
         mlflow.log_artifacts(model_dir, DEFAULT_MODEL_PATH)
 
+    def log_custom_model(self, model_dir=None, image="", command="", args=""):
+        """
+        Upload model to artifact storage.
+        This method is used to upload model for custom model type.
+
+        :param model_dir: directory which contain serialized model
+        :param image: Docker image that will be used as predictor
```
Comment (pradithya) [inline python/sdk/merlin/model.py:924]: `image` should not be optional argument

File: python/sdk/merlin/fluent.py [hunk 7]
```diff
@@ -322,6 +322,21 @@ def log_model(model_dir):
     _check_active_model_version()
     _active_model_version.log_model(model_dir)  # type: ignore
 
+def log_custom_model(model_dir: str, image :str, command: str, args: str):
```
Comment (pradithya) [inline python/sdk/merlin/fluent.py:325]: The signature of this function must be consistent with `ModelVersion#log_custom_model`

File: api/client/model_custom_predictor.go (hunk 5 above)
Comment (tiopramayudi) [reply api/client/model_custom_predictor.go:16]: the purpose is different, `ArtifactUri` always there when we create new version, but it doesn't mean that `ArtifactUri` is exist in GCS. That's why I add new field `IsArtifactExist` to determine whether users are trying to upload their artifact to mlflow or not

**Review by pradithya — APPROVED** [review]
LGTM!

**Pull request merged**